go-controller/pkg/ovn/egressip.go

```diff
@@ -2442,13 +2485,78 @@ func (e *EgressIPController) deleteExternalGWPodSNATOps(ni util.NetInfo, ops []o
 	return ops, nil
 }
 
+// getLayerScopedNextHop determines the next hop for a given Node considering the network topology type
+func (e *EgressIPController) getLayerScopedNextHop(ni util.NetInfo, nodeName string, wantsIPv6 bool) (net.IP, error) {
+	if ni.TopologyType() == types.Layer3Topology {
+		return e.getGatewayRouterJoinIP(ni, nodeName, wantsIPv6)
+	}
+	if ni.TopologyType() == types.Layer2Topology {
+		// For L2, if the pod selected by an EgressIP and also that EgressIP is managed by this controller, then the next hop must be
+		// retrieved from the node annotation. If a node is managed by another controller, we must get the remote Node gateway router to switch port IP
+		// from the node annotation.
+		// FIXME: remove gathering the required information from a Node annotations as this approach does not scale
+		node, err := e.watchFactory.GetNode(nodeName)
+		if err != nil {
+			return nil, fmt.Errorf("faild to retrive Node %s: %w", nodeName, err)
+		}
+		localNode, err := e.getALocalZoneNodeName()
+		if err != nil {
+			return nil, err
+		}
+		// node manager by this controller
+		if localNode == nodeName {
+			nextHopIPs, err := util.ParseNodeL3GatewayAnnotation(node)
+			if err != nil {
+				if util.IsAnnotationNotSetError(err) {
+					// remote node may not have the annotation yet, suppress it
+					return nil, types.NewSuppressedError(err)
+				}
+				return nil, fmt.Errorf("failed to get the node %s L3 gateway annotation: %w", node.Name, err)
+			}
+			if len(nextHopIPs.NextHops) == 0 {
+				return nil, fmt.Errorf("annotation on the node %s had empty next hop IPs", node.Name)
+			}
+			return util.MatchFirstIPFamily(wantsIPv6, nextHopIPs.NextHops)
+		}
+		// node managed by a different controller
+		return e.getGatewayRouterSwitchIP(ni, node, wantsIPv6)
+	}
+	return nil, fmt.Errorf("unsupported network %s", ni.TopologyType())
+}
+
 func (e *EgressIPController) getGatewayRouterJoinIP(ni util.NetInfo, node string, wantsIPv6 bool) (net.IP, error) {
-	gatewayIPs, err := libovsdbutil.GetLRPAddrs(e.nbClient, types.GWRouterToJoinSwitchPrefix+ni.GetNetworkScopedGWRouterName(node))
+	return e.getRouterPortIP(types.GWRouterToJoinSwitchPrefix+ni.GetNetworkScopedGWRouterName(node), wantsIPv6)
+}
+
+func (e *EgressIPController) getGatewayRouterSwitchIP(ni util.NetInfo, node *kapi.Node, wantsIPv6 bool) (net.IP, error) {
```

Comment:
awkward name with "RouterSwitch" maybe getGatewayRouterJoinSubnetIP

Reply:
why do you have this instad of just using func (e *EgressIPController) getGatewayRouterJoinIP ? or why does that function exist?

there's also util functions available:

```
func ParseNodeGatewayRouterJoinNetwork(node *kapi.Node, netName string) (primaryIfAddrAnnotation, error) {
	var val primaryIfAddrAnnotation
	joinSubnetMap, err := parseJoinSubnetAnnotation(node.Annotations, OVNNodeGRLRPAddrs)
	if err != nil {
		return val, fmt.Errorf("unable to parse annotation %s on node %s: err %w",
			OVNNodeGRLRPAddrs, node.Name, err)
	}
	val, ok := joinSubnetMap[netName]
	if !ok {
		return val, newAnnotationNotSetError("unable to fetch annotation value on node %s for network %s",
			node.Name, netName)
	}
	return val, nil
}

// ParseNodeGatewayRouterJoinIPv4 returns the IPv4 address for the node's gateway router port
// stored in the 'OVNNodeGRLRPAddrs' annotation
func ParseNodeGatewayRouterJoinIPv4(node *kapi.Node, netName string) (net.IP, error) {
	primaryIfAddr, err := ParseNodeGatewayRouterJoinNetwork(node, netName)
	if err != nil {
		return nil, err
	}
	if primaryIfAddr.IPv4 == "" {
		return nil, fmt.Errorf("failed to find an IPv4 address for gateway route interface in node: %s, net: %s, "+
			"annotation values: %+v", node, netName, primaryIfAddr)
	}

	ip, _, err := net.ParseCIDR(primaryIfAddr.IPv4)
	if err != nil {
		return nil, fmt.Errorf("failed to parse gateway router IPv4 address %s, err: %w", primaryIfAddr.IPv4, err)
	}
	return ip, nil
}
```

Reply:
> why do you have this instad of just using func (e *EgressIPController) getGatewayRouterJoinIP ? or why does that function exist?

I removed that func. It didnt increase code readability imo. 
I am using `getRouterPortIP`  directly without wrapping it with `getGatewayRouterJoinIP`.

I am using the utils now. I didn't know they existed.  I created one for IPv6.